docs/features/wired-mesh: update to role-based configuration (#2584)
freifunk-gluon/gluon: docs/features/wired-mesh.rst

of the *Advanced settings* (if the package ``gluon-web-network`` is installed).

It is also possible to enable Mesh-on-WAN and Mesh-on-LAN by default by adding
the ``mesh`` role to the ``interfaces.*.default_roles`` options in your
:ref:`site.conf<user-site-interfaces>`.

Commandline
===========

Starting with release 2022.1, the wired network configuration is rebuilt from ``/etc/config/gluon``
upon each ``gluon-reconfigure``.
Therefore the network configuration is overwritten at least with every firmware upgrade.

Every interface has a list of roles assigned to it which can be ``client``, ``mesh`` or ``uplink``.

When the client role is assigned to an interface in combination with other roles 
(like 'client', 'mesh' in the Mesh-on-LAN example below), the other roles take
precedence, enabling mesh but not client in the previous example.

The setup/config-mode interface is every interface with the role ``client`` which makes removing
it from interfaces not only unnecessary, but generally unrecommended.

> In order to make persistent changes to the routers configuration it's necessary to:

Reviewer: router's

> In order to make persistent changes to the routers configuration it's necessary to:

Reviewer: ```suggestion
In order to make persistent changes to the router's configuration it's necessary to:
```

Author: nice, thanks

* change the sections in ``/etc/config/gluon`` e.g. using uci (see examples below)
* call ``gluon-reconfigure`` to re-generate ``/etc/config/network``
* apply the networking changes, either through executing ``service network restart`` or by performing a ``reboot``

Enable Mesh-on-WAN::

  uci add_list gluon.iface_wan.role='mesh'
  uci commit gluon

Disable Mesh-on-WAN::

  uci del_list gluon.iface_wan.role='mesh'
  uci commit gluon

Enable Mesh-on-LAN::

  uci add_list gluon.iface_lan.role='mesh'
  uci commit gluon

Disable Mesh-on-LAN::

  uci del_list gluon.iface_lan.role='mesh'
  uci commit gluon

For devices with a single interface, instead of `iface_lan` and `iface_wan` configuration is
done with `iface_single`.

Enable Mesh-on-Single::

  uci add_list gluon.iface_single.role='mesh'
  uci commit gluon

Disable Mesh-on-Single::

  uci del_list gluon.iface_single.role='mesh'
  uci commit gluon

Furthermore it is possible to make use of 802.1Q VLAN.

> The following statements would create a VLAN with id 8 on ``eth0`` and join the mesh network with it::

Reviewer: double :: at the end of line?

> The following statements would create a VLAN with id 8 on ``eth0`` and join the mesh network with it::

Reviewer: ```suggestion
The following statements would create a VLAN with id 8 on ``eth0`` and join the mesh network with it:
```

Author: The double-double colon is necessary for the following code block to be formatted properly. Can't commit this.

  uci set gluon.iface_lan_vlan8=interface
  uci set gluon.iface_lan_vlan8.name='eth0.8'
  uci add_list gluon.iface_lan_vlan8.role='mesh'
  uci commit gluon

Other VLAN-interfaces could be configured on the same parent interface in order to have

> all three roles avaliable on ``eth0`` without having them interfere with each other.

Reviewer: ```suggestion
all three roles available on ``eth0`` without having them interfere with each other.
```

Author: nice catch, thanks.

Author: 714e2e3439eaae202a4cf78a520a30f8c02f2605

> This feature comes in especially handy for the persistent configuration of virtual machines

Reviewer: ```suggestion
This feature comes in handy especially for the persistent configuration of virtual machines
```
but i'm not too sure about this change

Author: They are both valid english;
I meant what I wrote, the changed variant changes the meaning of the sentence and feels like a weird shoulder pat to ourselves, to me.

as offloader for bigger installations.

> A ``reboot`` is not sufficient to apply altered configuration, a previous ``gluon-reconfigure``

Reviewer: ```suggestion
A ``reboot`` is not sufficient to apply an altered configuration, a prior ``gluon-reconfigure``
```

Author: reworded it a little:

A ``reboot`` is not sufficient to apply an altered configuration, calling ``gluon-reconfigure``
before is mandatory in order for changes to take effect.

Commandline
===========

Since version 2022.1, the wired network configuration is rebuilt from ``/etc/config/gluon`` on each ``gluon-reconfigure``, and therefore overwritten at least on every update of the node.
There are currently three roles available for interfaces: ``client``, ``mesh`` and ``uplink``.

> Be careful: the ``client`` role must not be mixed with the other two.

Author: You are right; I hope it's clearer now.

In order to make persistent changes to the routers configuration it's therefore neccessary to:

* change the sections in ``/etc/config/gluon`` (see examples below)
* call ``gluon-reconfigure`` to regenerate ``/etc/config/network``
* restart networking using either ``/etc/init.d/network restart`` or cleaner ``reboot`` the device

Enable Mesh-on-WAN::

  config interface 'iface_wan'
	option name '/wan'
	list role 'mesh'
	list role 'uplink'

Disable Mesh-on-WAN::

  config interface 'iface_wan'
	option name '/wan'
	list role 'uplink'

Enable Mesh-on-LAN::

  config interface 'iface_lan'
	option name '/lan'
	list role 'mesh'

Disable Mesh-on-LAN::

  config interface 'iface_lan'
	option name '/lan'

> 	list role 'client'

Author: I thought showing the config file would be easier to understand, but looking at the (slightly repetitive) mesh section now, just looks like a show of, of how clean the configuration can become ;)
Thanks for the hint.
I think this is resolved in 8e736b8c7d920a995ba4cc582b740f9cb10fdfcd.

It is furthermore possible to make use of VLANs.
The following section would create a VLAN with id 8 on ``eth0`` and emit the mesh network on it::

  config interface 'iface_lan_vlan8'
	option name 'eth0.8'
	list role 'mesh'

Other VLAN-interfaces could be configured on the same parent interface in order to have all three roles avaliable on ``eth0`` without having them interfere with each other.
This feature comes in especially handy in the persistent configuration of virtual machines as offloader for bigger installations.

> 

Reviewer 2: ```suggestion

It is possible to configure multiple interfaces with the same role, but currently all those interfaces will be bridged without any isolation. 
Mesh nodes that are connected to one of the mesh interfaces of a node can see each other directly.

```

sadly it doesn't look like #2484 (or something like it) will be merged in time for 2022.1

but idk, maybe it's better to leave this "disclaimer" away

Author: I agree on the other two points, but looking at the fact that we intend to change the behaviour, I'd avoid explicitly documenting that there's no isolation.

Commandline
===========

Since version 2022.1, the wired network configuration is rebuilt from ``/etc/config/gluon`` on each ``gluon-reconfigure``, and therefore overwritten at least on every firmwareupgrade.
Every interface has a list of roles assigned to it.
There are currently three roles available: ``client``, ``mesh`` and ``uplink``.

Be careful: When the client role is assigned to an interface at the same time as other roles (like 'client', 'mesh' in the Mesh-on-LAN example below), the other roles take precedence (enabling mesh, but not client in said example).

The setup/config-mode interface is every interface with the role ``client``, which makes removing it from interfaces not only unnecessary, but generally unrecommended.

In order to make persistent changes to the routers configuration it's necessary to:

* change the sections in ``/etc/config/gluon`` e.g. using uci (see examples below)

> * call ``gluon-reconfigure`` to regenerate ``/etc/config/network`` afterwords

Author: applied in 9ff834dd1f5a36d9f916bd1aa0527b5372ed2f82

> * restart networking using either ``/etc/init.d/network restart`` or cleaner ``reboot`` the device

Author: applied in 9ff834dd1f5a36d9f916bd1aa0527b5372ed2f82

> * apply the networking changes, either through executing ``service network restart`` or by performing a ``reboot`` of the device

Author: removed in 4a2c4dbc675f37c9fcabe5e30fd9e75286a9ee98

Enable Mesh-on-WAN::

  uci add_list gluon.iface_wan.role='mesh'
  uci commit gluon

Disable Mesh-on-WAN::

  uci del_list gluon.iface_wan.role='mesh'
  uci commit gluon

Enable Mesh-on-LAN::

  uci add_list gluon.iface_lan.role='mesh'
  uci commit gluon

Disable Mesh-on-LAN::

  uci del_list gluon.iface_lan.role='mesh'
  uci commit gluon

> 

Reviewer 2: ```suggestion

For devices with a single Network Port the commands are:

```

Author: I added them lower case.

Author: 4a2c4dbc675f37c9fcabe5e30fd9e75286a9ee98

Author: I'm not sure I like that too much.
But the idea very. Maybe I'd reword it to get rid of the colon, as it doubles in the next line.

`For devices with a single network port, instead of `iface_lan` and `iface_wan` configuration is done on `iface_single` as it's more specific.

Author: > For single-port-devices, instead of `iface_lan` and `iface_wan` configuration is done in `iface_single`.

Author: f988863c4403d6f8d809cbf9cfdd86f70c0db904

Enable Mesh-on-Single::

  uci add_list gluon.iface_single.role='mesh'
  uci commit gluon

Disable Mesh-on-Single::

  uci del_list gluon.iface_single.role='mesh'
  uci commit gluon

Furthermore it is possible to make use of VLANs.
The following statements would create a VLAN with id 8 on ``eth0`` and emit the mesh network on it::

  uci set gluon.iface_lan_vlan8=interface
  uci set gluon.iface_lan_vlan8.name='eth0.8'
  uci add_list gluon.iface_lan_vlan8.role='mesh'
  uci commit gluon

Other VLAN-interfaces could be configured on the same parent interface in order to have all three roles avaliable on ``eth0`` without having them interfere with each other.

> This feature comes in especially handy in the persistent configuration of virtual machines as offloader for bigger installations.

Author: applied in 9ff834dd1f5a36d9f916bd1aa0527b5372ed2f82

> It is not enough to ``reboot`` the device, calling ``gluon-reconfigure`` before is mandatory in order for the changes to take effect.

Reviewer 2: It's a bit redundant, but maybe it helps people, because it is quite important to know, idk

Author: I'm not sure about it either.
I felt like it wasn't stressed enough before.
